heat/cluster/spectral.py

```diff
@@ -0,0 +1,309 @@
+        Similarity metrics for adjacency calculations are supported via spatial.distance. The eigenvalues and eigenvectors are computed by reducing the Laplacian via lanczos iterations and using the torch eigenvalue solver on this smaller matrix. If other eigenvalue decompostion methods are supported, this will be expanded.
+
+
+        Parameters
+        ----------
+        X : ht.DNDarray, shape=(n_samples, n_features)
+            Training instances to cluster.
+        """
+        # input sanitation
+        if not isinstance(X, ht.DNDarray):
+            raise ValueError("input needs to be a ht.DNDarray, but was {}".format(type(X)))
+        if X.split is not None and X.split != 0:
+            raise NotImplementedError("Not implemented for other splitting-axes")
+        # 1. Calculation of Adjacency Matrix
+        if self.metric == "rbf":
+            sig = math.sqrt(1 / (2 * self.gamma))
+            S = ht.spatial.rbf(X, sigma=sig, quadratic_expansion=True)
+
+        elif self.metric == "euclidean":
+            S = ht.spatial.cdist(X)
+        else:
+            raise NotImplementedError("Other kernels currently not implemented")
+
+        # 2. Calculation of Laplacian
+        if self.laplacian == "eNeighborhood":
+            if self.epsilon[1] == "upper":
+                L = laplacian(S, norm=self.normalize, mode="eNeighbour", upper=self.epsilon[0])
+            elif self.epsilon[1] == "lower":
+                L = laplacian(S, norm=self.normalize, mode="eNeighbour", lower=self.epsilon[0])
+            else:
+                raise ValueError(
+                    "Boundary needs to be 'upper' or 'lower' and threshold needs to be set, if laplacian = eNeighborhood"
+                )
+
+        elif self.laplacian == "fully_connected":
+            L = laplacian(S, norm=self.normalize, mode="fc")
+        else:
+            raise NotImplementedError("Other approaches currently not implemented")
+
+        # 3. Eigenvalue and -vector calculation via Lanczos Algorithm
+        v0 = ht.ones((L.shape[0],), dtype=L.dtype, split=L.split, device=L.device) / math.sqrt(
+            L.shape[0]
+        )
+
+        V, T = ht.lanczos(L, self.n_lanczos, v0)
+        # 4. Calculate and Sort Eigenvalues and Eigenvectors of tridiagonal matrix T
+        eval, evec = torch.eig(T._DNDarray__array, eigenvectors=True)
+        # If x is an Eigenvector of T, then y = V@x is the corresponding Eigenvector of L
+        eval, idx = torch.sort(eval[:, 0], dim=0)
+        eigenvalues = ht.array(eval)
+        eigenvectors = ht.matmul(V, ht.array(evec))[:, idx]
+
+        # 5. Find the spectral gap, if number of clusters is not defined from the outside
+        if self.n_clusters is None:
+            temp = np.diff(eigenvalues.numpy())
```

Comment:
can we at least use torch here?

Reply:
The problem will be the line below 
```
self.n_clusters = np.where(diff == diff.max())[0][0] + 1
```
torch.where works as substitution on tensors, and does not yield a direct index, which is needed here. I thought since this is a small array for an internal check, using numpy wouldn't be a problem (similar to ``` numpy.full ``` in ``` counts_displs_shape ```.
Otherwise we need to reimplement all those small utility functions

Reply:
Cannot reproduce. Torch is doing the exact same thing for me as numpy is. Am I missing something here>

Reply:
Turns out, HeAT can do this, I was informed that ht.where does not return indices, but apparently it does. So ok, it's changed

Reply:
Needs some extra hacks, though